sycl/include/CL/sycl/kernel_bundle.hpp

```diff
@@ -0,0 +1,572 @@
+  backend get_backend() const noexcept;
+
+  /// Returns the context associated with the kernel_bundle
+  context get_context() const noexcept;
+
+  /// Returns devices associated with the kernel_bundle
+  std::vector<device> get_devices() const noexcept;
+
+  /// Returns true if the kernel_bundle contains the kernel identified by
+  /// kernel_id passed
+  bool has_kernel(const kernel_id &KernelID) const noexcept;
+
+  /// Returns true if the kernel_bundle contains the kernel identified by
+  /// kernel_id passed and if this kernel is compatible with the device
+  /// specified
+  bool has_kernel(const kernel_id &KernelID, const device &Dev) const noexcept;
+
+  /// Returns a vector of kernel_id's that contained in the kernel_bundle
+  std::vector<kernel_id> get_kernel_ids() const;
+
+  /// Returns true if the kernel_bundle contains at least one device image which
+  /// uses specialization constants
+  bool contains_specialization_constants() const noexcept;
+
+  /// Returns true if all specialization constants which are used in the
+  /// kernel_bundle are "native specialization constants in all device images
+  bool native_specialization_constant() const noexcept;
+
+protected:
+  // Returns true if the kernel_bundle has the specialization constant with
+  // specified ID
+  bool has_specialization_constant(unsigned int SpecID) const noexcept;
+
+  // Sets the specialization constant with specified ID to the value pointed by
+  // Value + ValueSize
+  void set_specialization_constant(unsigned int SpecID, void *Value,
+                                   size_t ValueSize);
+
+  // Returns pointer to the value of the specialization constant with specified
+  // ID
+  void *get_specialization_constant(unsigned int SpecID) const;
+
+  // Returns a kernel object which represents the kernel identified by
+  // kernel_id passed
+  kernel get_kernel(const kernel_id &KernelID) const;
+
+  // Returns an iterator to the first device image kernel_bundle contains
+  kernel_bundle_plain *begin() const;
+
+  // Returns an iterator to the last device image kernel_bundle contains
+  kernel_bundle_plain *end() const;
+
+  detail::KernelBundleImplPtr impl;
+};
+
+} // namespace detail
+
+/// The kernel_bundle class represents collection of device images in a
+/// particular state
+///
+/// \ingroup sycl_api
+template <bundle_state State>
+class kernel_bundle : public detail::kernel_bundle_plain {
+public:
+  using device_image_iterator = device_image<State> *;
+
+  kernel_bundle() = delete;
+
+  // some required methods come from the base class
+
+  /// Returns a kernel object which represents the kernel identified by
+  /// kernel_id passed
+  template <bundle_state _State = State,
+            typename = detail::enable_if_t<_State == bundle_state::executable>>
+  kernel get_kernel(const kernel_id &KernelID) const {
+    return detail::kernel_bundle_plain::get_kernel(KernelID);
+  }
+
+#if __cplusplus > 201402L
```

Comment:
This guard is here to because the libsycl.so can be compiled with C++ <= 14.
The ifdef should not be a problem since this functions are not supposed to be used by the libsycl.so